samza-core/src/main/java/org/apache/samza/storage/StoreBackupEngine.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.conf.*;
+
+// Class to hold common methods and variables for backup storage and restore process
+public class StoreBackupEngine {
+  protected static final Logger LOGGER = LoggerFactory.getLogger(StoreBackupEngine.class);
+  protected FileSystem fs;
+  protected Path hdfsPath;
+  protected Path dbPath;
+  public StoreBackupMetrics metrics;
+
+  public StoreBackupEngine(Config storeConfig) {
+    metrics = new StoreBackupMetrics("backup", new MetricsRegistryMap("registrymap"));
+
+    // Setup fs configuration settings
+    Configuration conf = new Configuration();
+    // If core site path file is specified
+    if (storeConfig.containsKey("hdfs.coreSitePath")){
+      Path coreSitePath = new Path(storeConfig.get("hdfs.coreSitePath"));
+      conf.addResource(coreSitePath);
+    } else {
+      // Otherwise, set a default localhost port for fs.defaultFS - hdfs://localhost:9000
+      conf.set("fs.default.name", "hdfs://localhost:9000");
+    }
+
+    // Initialize path variables
+    // Method: config.get(key, defaultVal)
+    dbPath = new Path(storeConfig.get("rocksdb.dbDir", System.getProperty("java.io.tmpdir") + "/DB"));
+
+    // Set where to store backup files on hdfs
+    // TODO: change default value to something else for backup since obviously below default only works for me
+    hdfsPath = new Path(storeConfig.get("hdfs.backupDir", "/User/eripan/backupTest/"));
+
+    // Possible IOException when using filesystem api
+    try {
+      fs = FileSystem.get(conf);
+
+      // Create directory for hdfs backup location
+      // No need to check if localPath is dir since it's checked when rocksdb is created
+      if (fs.exists(hdfsPath)) {
+        if (fs.isFile(hdfsPath)) {
+          throw new SamzaException("Error creating backup in hdfs: Path is already a file");
+        }
+      } else {
+        fs.mkdirs(hdfsPath);
+      }
+
+    } catch(IOException e) {
+      throw new SamzaException("Failed to create hdfs filesystem: " + e);
+    }
+  }
+
+  // Method to restore rocksdb state from hdfs backup
+  public Boolean restoreBackup() throws IOException {
+    metrics.restores().inc();
+
+    // Check if db directory has already been created or not
+    File dbDir = new File(dbPath.toString());
+    if(!dbDir.exists()) {
+      dbDir.mkdirs();
+    }
+
+    // Variables to keep track of statistics
+    long startTime = System.currentTimeMillis();
+    long totalBytesCopied = 0;
+    long totalBytesDeleted = 0;
+
+    // Find what files to copy from backup compare to what's in Rocksdb
+    ArrayList<FileStatus> backupFilesToCopy = getNewFilesInBackupToLocal(dbPath);
+    for (FileStatus backupFile : backupFilesToCopy) {
+      LOGGER.info("Found file: " + backupFile.getPath().getName());
+      totalBytesCopied += backupFile.getLen();
+      // Method: copyToLocalFile(hdfsFilePath, localFilePath)
+      fs.copyToLocalFile(backupFile.getPath(), dbPath);
```

Comment:
Can this throw an exception? What happens if this fails part-way during the copy? Is the store still usable? Maybe we should make it such that the MANIFEST etc. are copied last?